[AIE2P] Add alternative calling convention for library calls

konstantinschwarz opened the pull request:
In AIEs default calling convention, all vector registers are caller saved. However, most compiler-rt builtin library functions do not use vector registers. To avoid having to spill vectors to memory in the caller function, we can attach a different calling convention to these library functions.

I manually inspected each library function where we apply the alternative calling convention to make sure none of them actually use vector registers. To make this actually save and robust, we should also add a custom calling convention attribute to the frontend that ensures vector registers are actually saved by the callee.

konstantinschwarz commented:
This saves up to 512 bytes of stack (ArgMin/ArgMax), and some very minor code size savings.

<img width="4400" height="1800" alt="Core_PMSize_absolute" src="https://github.com/user-attachments/assets/a12dba54-da33-4ee1-9d5c-71e6545ab0a0" />
<img width="4400" height="1800" alt="Core_StackSize_absolute" src="https://github.com/user-attachments/assets/404b7821-e714-43ac-a361-87442c6b0fb8" />

andcarminati commented on llvm/lib/Target/AIE/aie2p/AIE2PISelLowering.cpp:
```diff
@@ -21,9 +21,34 @@ using namespace llvm;
+    setLibcallCallingConv(RTLIB::SREM_I32, CallingConv::AIE_PreserveAll_Vec);
+    setLibcallCallingConv(RTLIB::SREM_I64, CallingConv::AIE_PreserveAll_Vec);
+    setLibcallCallingConv(RTLIB::UREM_I32, CallingConv::AIE_PreserveAll_Vec);
+    setLibcallCallingConv(RTLIB::UREM_I64, CallingConv::AIE_PreserveAll_Vec);
```
I see some other libcalls in the LegalizerHelper. Are these only the verified ones?

konstantinschwarz replied:
Yes, these are only the ones I verified. Others could likely be included as well, but I wanted to keep this limited for now

martien-de-jong commented on llvm/lib/Target/AIE/aie2p/AIE2PCallingConv.td:
```diff
@@ -219,5 +219,12 @@ def CC_AIE2P : CallingConv<[
     : CalleeSavedRegs<(add lr, r8, r9, r10, r11, r12, r13, r14, r15, p6, p7)>;
 
+def CSR_AIE2P_Vec
+    : CalleeSavedRegs<(add lr, r8, r9, r10, r11, r12, r13, r14, r15, p6, p7,
```
nit: can we write this as an extension to CSR_AIE2P?

martien-de-jong replied:
e.g. factoring using a named common set?

martien-de-jong commented on llvm/lib/Target/AIE/aie2p/AIE2PISelLowering.cpp:
```diff
@@ -22,9 +22,34 @@ using namespace llvm;
     : AIEBaseTargetLowering(TM, STI) {
+
+  if (VecCCLibcalls) {
+    setLibcallCallingConv(RTLIB::SINTTOFP_I32_F32,
```
nit: perhaps make a data-driven routine running on an array of function ids?

konstantinschwarz committed:
[AIE2P][AIE2PS] Add alternative calling convention for library calls

In AIEs default calling convention, all vector registers are caller saved.
However, most compiler-rt builtin library functions do not use vector registers.
To avoid having to spill vectors to memory in the caller function,
we can attach a different calling convention to these library functions.

konstantinschwarz commented:
Rebased and added support for aie2ps. @F-Stuckmann can you approve again?
